File: example/wallet-as/assembly/procedures.ts
import {
  Oid,
  ResultSet,
  SqlStmt,
  Value,
  ValueList,
} from "./mududb_binding";
import { witCommand, witQuery } from "./mududb_wit";

function command(id: Oid, sql: string, values: ValueList = new ValueList()): i64 {
  return <i64>witCommand(id, new SqlStmt(sql).raw, values.raw).unwrap();
}

function queryBalance(id: Oid, userId: i64): i64 {
  const params = new ValueList();
  params.bind(0, Value.int64(userId));

  const rows = new ResultSet(witQuery(
    id,
    new SqlStmt("SELECT balance FROM wallets WHERE user_id = ?").raw,
    params.raw,
  ).unwrap());

  if (!rows.next()) {
    throw new Error("wallet not found");
  }
  return rows.currentRow().valueByName("balance").asInt64();
}

function setBalance(id: Oid, userId: i64, balance: i64): i64 {
  const params = new ValueList();
  params.bind(0, Value.int64(balance));
  params.bind(1, Value.int64(userId));

  const updated = command(
    id,
    "UPDATE wallets SET balance = ? WHERE user_id = ?",
    params,
  );
  if (updated != 1) {
    throw new Error("wallet update failed");
  }
  return balance;
}

/**mudu-proc*/
export function create_user(id: Oid, user_id: i64, name: string, email: string): i64 {
  const userParams = new ValueList();
  userParams.bind(0, Value.int64(user_id));
  userParams.bind(1, Value.text(name));
  userParams.bind(2, Value.text(email));
  userParams.bind(3, Value.int64(0));
  userParams.bind(4, Value.int64(0));
  const users = command(
    id,
    "INSERT INTO users (user_id, name, email, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
    userParams,
  );
  if (users != 1) {
    throw new Error("create user failed");
  }

  const walletParams = new ValueList();
  walletParams.bind(0, Value.int64(user_id));
  walletParams.bind(1, Value.int64(0));
  walletParams.bind(2, Value.int64(0));
  const wallets = command(
    id,
    "INSERT INTO wallets (user_id, balance, updated_at) VALUES (?, ?, ?)",
    walletParams,
  );
  if (wallets != 1) {
    throw new Error("create wallet failed");
  }
  return user_id;
}

/**mudu-proc*/
export function deposit(id: Oid, user_id: i64, amount: i64): i64 {
  if (amount <= 0) {
    throw new Error("amount must be positive");
  }

  const balance = queryBalance(id, user_id);
  return setBalance(id, user_id, balance + amount);
}

/**mudu-proc*/
export function withdraw(id: Oid, user_id: i64, amount: i64): i64 {
  if (amount <= 0) {
    throw new Error("amount must be positive");
  }

  const current = queryBalance(id, user_id);
  if (current < amount) {
    throw new Error("insufficient funds");
  }
  return setBalance(id, user_id, current - amount);
}

/**mudu-proc*/
export function transfer_funds(
  id: Oid,
  from_user_id: i64,
  to_user_id: i64,
  amount: i64,
): i64 {
  if (amount <= 0) {
    throw new Error("amount must be positive");
  }
  if (from_user_id == to_user_id) {
    throw new Error("cannot transfer to self");
  }

  const currentFrom = queryBalance(id, from_user_id);
  const currentTo = queryBalance(id, to_user_id);
  if (currentFrom < amount) {
    throw new Error("insufficient funds");
  }

  const newFrom = setBalance(id, from_user_id, currentFrom - amount);
  setBalance(id, to_user_id, currentTo + amount);
  return newFrom;
}

/**mudu-proc*/
export function balance(id: Oid, user_id: i64): i64 {
  return queryBalance(id, user_id);
}

File: crates/sdk/example/wallet-cs/src/Procedures.cs
#nullable enable

namespace WalletCs;

/// <summary>
/// Wallet business logic, mirroring the AssemblyScript wallet-as procedures
/// (<c>crates/sdk/example/wallet-as/assembly/procedures.ts</c>) one to one:
/// every procedure takes the session OID from the procedure parameter and
/// returns an <c>i64</c> (the affected user id or the resulting balance).
/// SQL goes through <see cref="MuduSys"/>, which frames SyscallPayload v1
/// (MSSP) messages over the <c>mududb:api/system</c> byte pipe.
/// </summary>
internal static class Procedures
{
    // mudu-proc
    public static long CreateUser(MuduOid session, long userId, string name, string email)
    {
        var users = MuduSys.Command(session,
            "INSERT INTO users (user_id, name, email, created_at, updated_at) VALUES (?, ?, ?, ?, ?)",
            userId, name, email, 0L, 0L);
        if (users != 1)
        {
            throw new WalletCsException(ErrorCodes.DomainViolation, "create user failed");
        }

        var wallets = MuduSys.Command(session,
            "INSERT INTO wallets (user_id, balance, updated_at) VALUES (?, ?, ?)",
            userId, 0L, 0L);
        if (wallets != 1)
        {
            throw new WalletCsException(ErrorCodes.DomainViolation, "create wallet failed");
        }

        return userId;
    }

    // mudu-proc
    public static long Deposit(MuduOid session, long userId, long amount)
    {
        if (amount <= 0)
        {
            throw new WalletCsException(ErrorCodes.InvalidArgument, "amount must be positive");
        }

        var balance = QueryBalance(session, userId);
        return SetBalance(session, userId, balance + amount);
    }

    // mudu-proc
    public static long Withdraw(MuduOid session, long userId, long amount)
    {
        if (amount <= 0)
        {
            throw new WalletCsException(ErrorCodes.InvalidArgument, "amount must be positive");
        }

        var balance = QueryBalance(session, userId);
        if (balance < amount)
        {
            throw new WalletCsException(ErrorCodes.DomainViolation, "insufficient funds");
        }

        return SetBalance(session, userId, balance - amount);
    }

    // mudu-proc
    public static long TransferFunds(MuduOid session, long fromUserId, long toUserId, long amount)
    {
        if (amount <= 0)
        {
            throw new WalletCsException(ErrorCodes.InvalidArgument, "amount must be positive");
        }

        if (fromUserId == toUserId)
        {
            throw new WalletCsException(ErrorCodes.InvalidArgument, "cannot transfer to self");
        }

        var fromBalance = QueryBalance(session, fromUserId);
        var toBalance = QueryBalance(session, toUserId);
        if (fromBalance < amount)
        {
            throw new WalletCsException(ErrorCodes.DomainViolation, "insufficient funds");
        }

        var newFromBalance = SetBalance(session, fromUserId, fromBalance - amount);
        SetBalance(session, toUserId, toBalance + amount);
        return newFromBalance;
    }

    // mudu-proc
    public static long Balance(MuduOid session, long userId)
    {
        return QueryBalance(session, userId);
    }

    private static long QueryBalance(MuduOid session, long userId)
    {
        var rows = MuduSys.Query(session, "SELECT balance FROM wallets WHERE user_id = ?", userId);
        if (rows.Count == 0 || rows[0][0] is not long balance)
        {
            throw new WalletCsException(ErrorCodes.EntityNotFound, "wallet not found");
        }

        return balance;
    }

    private static long SetBalance(MuduOid session, long userId, long balance)
    {
        var updated = MuduSys.Command(session,
            "UPDATE wallets SET balance = ? WHERE user_id = ?",
            balance, userId);
        if (updated != 1)
        {
            throw new WalletCsException(ErrorCodes.DomainViolation, "wallet update failed");
        }

        return balance;
    }
}

/// <summary>
/// Numeric mirrors of the host's <c>mudu::error::ErrorCode</c> discriminants
/// carried in the <c>UniError</c> result arm.
/// </summary>
internal static class ErrorCodes
{
    public const uint EntityNotFound = 50009;
    public const uint DomainViolation = 50017;
    public const uint InvalidArgument = 50029;
    public const uint Internal = 50000;
}

/// <summary>
/// A domain error raised by a wallet procedure; the byte-pipe dispatcher
/// encodes it as the <c>{1: UniError}</c> arm of the procedure result.
/// </summary>
internal sealed class WalletCsException(uint code, string message)
    : global::System.Exception(message)
{
    public uint Code { get; } = code;
}
